File: ling_chat/core/ai_service/proactive_system/strategy_dispatcher.py
import os
import random
from datetime import datetime
from typing import Optional

from ling_chat.core.ai_service.game_system.game_status import GameStatus
from ling_chat.core.ai_service.proactive_system.type import PerceptionResult, UserState
from ling_chat.core.llm_providers.manager import LLMManager
from ling_chat.core.pic_analyzer import DesktopAnalyzer
from ling_chat.schemas.schedule_settings import TodoItem, UserScheduleSettings


class StrategyDispatcher:
    def __init__(self, game_status: GameStatus, settings: UserScheduleSettings):
        self.game_status = game_status
        self.settings = settings
        self.pending_topics = []
        self.llm_helper = LLMManager()  # 辅助LLM
        self.desktop_analyzer = DesktopAnalyzer()  # 桌面分析器

    async def get_proactive_prompt(self, perception: PerceptionResult) -> Optional[str]:
        """
        生成主动对话的 Prompt。
        优先顺序: ImportantDay (Only once) > Todo > Scene/Topic > Screen Observation
        """

        today_str = datetime.now().strftime("%m-%d")

        # 1. 检查 ImportantDay (如果是今天且未触发过)
        enable_important_day = (
            os.getenv("ENABLE_IMPORTANT_DAY_REMINDER", "false").lower() == "true"
        )
        if enable_important_day and self.settings.importantDays:
            last_talk_date = (
                # last_dialog_time 理论上在 chat_history.py 加载后已经是 datetime，
                # 但为防止极端情况（上层未初始化或上游改动），用 isinstance 兜底。
                self.game_status.last_dialog_time.strftime("%m-%d")
                if isinstance(self.game_status.last_dialog_time, datetime)
                else (str(self.game_status.last_dialog_time) if self.game_status.last_dialog_time else "")
            )
            if last_talk_date != today_str:
                for day in self.settings.importantDays:
                    if day.date.endswith(today_str):
                        return f"{{今天是特殊的一天：{day.title}，{day.desc}。可以和{self.game_status.player.user_name}聊聊哦}}"

        # 2. 随机模式选择，根据启用状态动态构建候选列表
        # 基础权重: TODO(10), TOPIC(60), SCREEN(30)
        # 禁用某个模式后，random.choices 会自动对剩余权重做归一化
        modes = []
        weights = []

        # 初始化权重
        todo_weight = None
        topic_weight = None
        screen_weight = None

        # 优先从环境变量读取权重
        try:
            todo_weight = int(os.getenv("TODO_WEIGHT", "-1"))
            topic_weight = int(os.getenv("TOPIC_WEIGHT", "-1"))
            screen_weight = int(os.getenv("SCREEN_WEIGHT", "-1"))
        except ValueError:
            pass

        # 根据用户状态动态设置权重（仅在环境变量未设置时）
        if todo_weight is None or todo_weight < 0:
            todo_weight = 60 if perception.state == UserState.WORK else 10

        if topic_weight is None or topic_weight < 0:
            topic_weight = 80 if perception.state == UserState.IDLE else 60

        if screen_weight is None or screen_weight < 0:
            screen_weight = 60 if perception.state == UserState.GAME else 30

        # 添加模式和对应的权重
        if os.getenv("ENABLE_TODO_PRECEPTION", "false").lower() == "true":
            modes.append("TODO")
            weights.append(todo_weight)

        if os.getenv("ENABLE_TOPIC_CREATER", "false").lower() == "true":
            modes.append("TOPIC")
            weights.append(topic_weight)

        if os.getenv("ENABLE_VISUAL_PRECEPTION", "true").lower() == "true":
            modes.append("SCREEN")
            weights.append(screen_weight)

        if not modes:
            return None

        mode = random.choices(modes, weights=weights)[0]

        if mode == "TODO":
            prompt = self._get_todo_prompt()
            if prompt:
                return prompt
            # 没有 Todo 时降级到 TOPIC（如果启用）
            if os.getenv("ENABLE_TOPIC_CREATER", "false").lower() == "true":
                return await self._get_topic_prompt()
            return None

        if mode == "SCREEN":
            analyze_prompt = "你是一个图像信息转述者，你将需要把你看到的画面描述给另一个AI让他理解用户的图片内容。用户开放了那个AI的自主窥屏功能，请获取桌面画面中的重点内容，用200字描述主体部分即可。如果你看到一个聊天窗口，有角色的立绘和对话框，不要描述这部分，只描述桌面上的其他内容。因为那部分是玩家与AI的聊天窗口。"
            analyze_info = await self.desktop_analyzer.analyze_desktop(analyze_prompt)
            ai_name = (
                self.game_status.current_character.display_name
                if self.game_status.current_character
                else "你"
            )
            return f"{{ {ai_name} 偷看了一眼 {self.game_status.player.user_name} 的电脑桌面信息: {analyze_info} }}"

        # TOPIC
        return await self._get_topic_prompt()

    def _get_todo_prompt(self) -> Optional[str]:
        """随机抽取未完成且重要的 Todo"""
        if not self.settings.todoGroups:
            return None

        candidates: list[TodoItem] = []
        for _, group in self.settings.todoGroups.items():
            for todo in group.todos:
                if not todo.completed and todo.priority >= 1:
                    candidates.append(todo)

        if not candidates:
            return None

        selected = random.choice(candidates)
        return f"{{你想起来{self.game_status.player.user_name}有一个未完成的任务：'{selected.text} '。提醒一下吧？}}"

    async def _get_topic_prompt(self) -> str:
        # TODO： 这里应该发送上下文给一个专门创造话题的 LLM
        """获取聊天话题"""
        if not self.pending_topics:
            # 话题枯竭，使用 LLM 生成补充
            # 这里是伪代码，实际需要调用 LLM 接口
            # new_topics = await self.llm_helper.generate_topics(self.game_status.memory)
            new_topics = ["天气", "最近的游戏", "休息一下"]
            self.pending_topics.extend(new_topics)

        _topic = self.pending_topics.pop(0) if self.pending_topics else "发呆"
        ai_name = (
            self.game_status.current_character.display_name
            if self.game_status.current_character
            else "你"
        )
        return f"{{ {ai_name} 想继续说话了}}"

File: src-tauri/src/ai_service/proactive_system/strategy_dispatcher.rs
use chrono::Local;
use rand::Rng;
use reqwest::Client;
use serde_json::Value;
use crate::ai_service::proactive_system::types::{
    UserScheduleSettings, PerceptionResult, UserState
};
use crate::ai_service::proactive_system::config::ProactiveConfig;
use crate::ai_service::game_system::game_status::GameStatus;

pub struct StrategyDispatcher;

impl StrategyDispatcher {
    pub fn new() -> Self {
        Self
    }

    /// 生成主动对话的 Prompt。
    /// 优先顺序: ImportantDay (每天仅一次) > Todo > Screen Observation > Topic
    pub async fn get_proactive_prompt(
        &self,
        game_status: &GameStatus,
        settings: &UserScheduleSettings,
        perception: &PerceptionResult,
        config: &ProactiveConfig,
    ) -> Option<String> {
        let now = Local::now();
        let today_str = now.format("%m-%d").to_string();

        // 1. 检查 ImportantDay (如果是今天且今天未触发过)
        if config.enable_important_day_reminder {
            if let Some(important_days) = &settings.important_days {
                let last_talk_date = game_status
                    .last_dialog_time
                    .map(|dt| dt.format("%m-%d").to_string())
                    .unwrap_or_default();

                if last_talk_date != today_str {
                    for day in important_days {
                        if day.date.ends_with(&today_str) {
                            let desc = day.desc.as_deref().unwrap_or("");
                            let char_name = game_status
                                .current_role_id
                                .and_then(|rid| game_status.role_manager.get_loaded(rid))
                                .and_then(|role| role.display_name.clone())
                                .unwrap_or_else(|| "小灵".to_string());
                            
                            log::info!("[StrategyDispatcher] Triggered important day reminder: {}", day.title);
                            return Some(format!(
                                "{{今天是特殊的一天：{}，{}。可以和{}聊聊哦}}",
                                day.title, desc, char_name
                            ));
                        }
                    }
                }
            }
        }

        // 2. 随机模式选择，根据启用状态动态构建候选列表
        let mut modes = Vec::new();
        let mut weights = Vec::new();

        // 获取权重（如果配置文件有设置，否则基于 UserState 动态决定）
        let mut todo_w = config.todo_weight;
        let mut topic_w = config.topic_weight;
        let mut screen_w = config.screen_weight;

        if todo_w <= 0.0 {
            todo_w = if perception.state == UserState::WORK { 60.0 } else { 10.0 };
        }
        if topic_w <= 0.0 {
            topic_w = if perception.state == UserState::IDLE { 80.0 } else { 60.0 };
        }
        if screen_w <= 0.0 {
            screen_w = if perception.state == UserState::GAME { 60.0 } else { 30.0 };
        }

        if config.enable_todo_perception {
            modes.push("TODO");
            weights.push(todo_w);
        }
        if config.enable_topic_creator {
            modes.push("TOPIC");
            weights.push(topic_w);
        }
        if config.enable_visual_perception {
            modes.push("SCREEN");
            weights.push(screen_w);
        }

        if modes.is_empty() {
            return None;
        }

        // 轮盘赌/加权随机选择
        let selected_mode = {
            let mut rng = rand::thread_rng();
            let total_weight: f64 = weights.iter().sum();
            let mut roll = rng.gen_range(0.0..total_weight);
            let mut selected = modes[0];
            for (i, &w) in weights.iter().enumerate() {
                roll -= w;
                if roll <= 0.0 {
                    selected = modes[i];
                    break;
                }
            }
            selected
        };

        log::info!(
            "[StrategyDispatcher] Selected proactive mode: {} (weights: TODO={:.1}, TOPIC={:.1}, SCREEN={:.1})",
            selected_mode, todo_w, topic_w, screen_w
        );

        match selected_mode {
            "TODO" => {
                if let Some(prompt) = self.get_todo_prompt(game_status, settings) {
                    return Some(prompt);
                }
                // 没有 Todo 时降级到 TOPIC
                if config.enable_topic_creator {
                    return Some(self.get_topic_prompt(game_status));
                }
                None
            }
            "SCREEN" => {
                if let Some(prompt) = self.get_screen_prompt(game_status, config).await {
                    return Some(prompt);
                }
                // SCREEN 抓取失败或接口失败时降级到 TOPIC
                if config.enable_topic_creator {
                    return Some(self.get_topic_prompt(game_status));
                }
                None
            }
            _ => Some(self.get_topic_prompt(game_status)),
        }
    }

    fn get_todo_prompt(
        &self,
        game_status: &GameStatus,
        settings: &UserScheduleSettings,
    ) -> Option<String> {
        let todo_groups = settings.todo_groups.as_ref()?;
        let mut candidates = Vec::new();

        for group in todo_groups.values() {
            for todo in &group.todos {
                if !todo.completed && todo.priority >= 1 {
                    candidates.push(todo);
                }
            }
        }

        if candidates.is_empty() {
            return None;
        }

        let mut rng = rand::thread_rng();
        let idx = rng.gen_range(0..candidates.len());
        let selected = candidates[idx];
        let user_name = &game_status.player.user_name;

        Some(format!(
            "{{你想起来{}有一个未完成的任务：'{} '。提醒一下吧？}}",
            user_name, selected.text
        ))
    }

    async fn get_screen_prompt(
        &self,
        game_status: &GameStatus,
        config: &ProactiveConfig,
    ) -> Option<String> {
        if config.vd_api_key.is_empty() {
            log::warn!("[StrategyDispatcher] VD_API_KEY is empty, skipping screenshot analysis.");
            return None;
        }

        // 截取桌面画面并压缩为 JPEG 字节
        let jpeg_bytes = match capture_screen_as_jpeg() {
            Some(bytes) => bytes,
            None => {
                log::error!("[StrategyDispatcher] Failed to capture desktop screenshot.");
                return None;
            }
        };

        let base64_image = base64::Engine::encode(&base64::prelude::BASE64_STANDARD, &jpeg_bytes);
        let image_url = format!("data:image/jpeg;base64,{}", base64_image);

        let analyze_prompt = "你是一个图像信息转述者，你将需要把你看到的画面描述给另一个AI让他理解用户的图片内容。用户开放了那个AI的自主窥屏功能，请获取桌面画面中的重点内容，用200字描述主体部分即可。如果你看到一个聊天窗口，有角色的立绘和对话框，不要描述这部分，只描述桌面上的其他内容。因为那部分是玩家与AI的聊天窗口。";

        log::info!("[StrategyDispatcher] Sending screenshot to Vision LLM ({}) for analysis...", config.vd_model);

        let client = Client::new();
        let payload = serde_json::json!({
            "model": &config.vd_model,
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {"type": "text", "text": analyze_prompt},
                        {"type": "image_url", "image_url": {"url": image_url}}
                    ]
                }
            ],
            "max_tokens": 512
        });

        let res = client.post(&format!("{}/chat/completions", config.vd_base_url))
            .bearer_auth(&config.vd_api_key)
            .json(&payload)
            .send()
            .await;

        match res {
            Ok(response) => {
                if response.status().is_success() {
                    if let Ok(json_res) = response.json::<Value>().await {
                        if let Some(content) = json_res["choices"][0]["message"]["content"].as_str() {
                            let user_name = &game_status.player.user_name;
                            let ai_name = game_status
                                .current_role_id
                                .and_then(|rid| game_status.role_manager.get_loaded(rid))
                                .and_then(|role| role.display_name.clone())
                                .unwrap_or_else(|| "你".to_string());
                            
                            log::info!("[StrategyDispatcher] Screenshot analysis success: {}", content);
                            return Some(format!(
                                "{{ {} 偷看了一眼 {} 的电脑桌面信息: {} }}",
                                ai_name, user_name, content
                            ));
                        }
                    }
                } else {
                    let err_text = response.text().await.unwrap_or_default();
                    log::error!("[StrategyDispatcher] VLM API returned error status: {}", err_text);
                }
            }
            Err(e) => {
                log::error!("[StrategyDispatcher] Failed to send request to VLM: {:?}", e);
            }
        }

        None
    }

    fn get_topic_prompt(&self, game_status: &GameStatus) -> String {
        let ai_name = game_status
            .current_role_id
            .and_then(|rid| game_status.role_manager.get_loaded(rid))
            .and_then(|role| role.display_name.clone())
            .unwrap_or_else(|| "你".to_string());
        
        format!("{{ {} 想继续说话了}}", ai_name)
    }
}

/// 捕获桌面并将其转换为 JPEG 格式的辅助函数。
fn capture_screen_as_jpeg() -> Option<Vec<u8>> {
    #[cfg(target_os = "windows")]
    {
        use windows::Win32::Graphics::Gdi::{
            GetDC, ReleaseDC, CreateCompatibleDC, CreateCompatibleBitmap, SelectObject,
            BitBlt, GetDIBits, DeleteDC, DeleteObject, SRCCOPY, BITMAPINFOHEADER, DIB_RGB_COLORS,
        };
        use windows::Win32::UI::WindowsAndMessaging::{GetSystemMetrics, SM_CXSCREEN, SM_CYSCREEN};

        unsafe {
            let hdc_screen = GetDC(None);
            if hdc_screen.is_invalid() {
                return None;
            }
            let w = GetSystemMetrics(SM_CXSCREEN);
            let h = GetSystemMetrics(SM_CYSCREEN);
            if w <= 0 || h <= 0 {
                ReleaseDC(None, hdc_screen);
                return None;
            }
            
            let hdc_mem = CreateCompatibleDC(Some(hdc_screen));
            let hbitmap = CreateCompatibleBitmap(hdc_screen, w, h);
            
            let old_obj = SelectObject(hdc_mem, hbitmap.into());
            let _ = BitBlt(hdc_mem, 0, 0, w, h, Some(hdc_screen), 0, 0, SRCCOPY);
            
            // setup BITMAPINFO
            let mut bmi = windows::Win32::Graphics::Gdi::BITMAPINFO::default();
            bmi.bmiHeader.biSize = std::mem::size_of::<BITMAPINFOHEADER>() as u32;
            bmi.bmiHeader.biWidth = w;
            bmi.bmiHeader.biHeight = -h; // Negative height for top-down bitmap
            bmi.bmiHeader.biPlanes = 1;
            bmi.bmiHeader.biBitCount = 24;
            bmi.bmiHeader.biCompression = 0; // BI_RGB
            
            let mut buffer = vec![0u8; (w * h * 3) as usize];
            
            let lines = GetDIBits(
                hdc_screen,
                hbitmap,
                0,
                h as u32,
                Some(buffer.as_mut_ptr() as *mut _),
                &mut bmi,
                DIB_RGB_COLORS,
            );
            
            // cleanup GDI objects
            SelectObject(hdc_mem, old_obj);
            let _ = DeleteObject(hbitmap.into());
            let _ = DeleteDC(hdc_mem);
            ReleaseDC(None, hdc_screen);
            
            if lines <= 0 {
                return None;
            }
            
            // Swap B and R to get RGB from BGR
            for chunk in buffer.chunks_exact_mut(3) {
                chunk.swap(0, 2);
            }
            
            // Compress image to 1024x768 for faster Vision API upload and fewer tokens
            let img = image::ImageBuffer::<image::Rgb<u8>, _>::from_raw(w as u32, h as u32, buffer)?;
            let dynamic_img = image::DynamicImage::ImageRgb8(img);
            let resized = dynamic_img.resize(1024, 768, image::imageops::FilterType::Triangle);
            
            let mut jpeg_bytes = Vec::new();
            let mut cursor = std::io::Cursor::new(&mut jpeg_bytes);
            let _ = resized.write_to(&mut cursor, image::ImageFormat::Jpeg);
            
            Some(jpeg_bytes)
        }
    }

    #[cfg(not(target_os = "windows"))]
    {
        None
    }
}
